bench/micro/Bench/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -63,22 +76,47 @@ benchLookups conf@Config{name} =
-            -- lookup keys. We use whnf here because the result is an unboxed
-            -- vector.
+            -- lookup keys. We use whnf here because the result is
           , env (pure $ bloomQueriesDefault blooms ks) $ \rkixs ->
               bench "Compact index search" $
                 whnfAppIO (\ks' -> indexSearches indexes kopsFiles ks' rkixs) ks
-            -- All prepped lookups are going to be used eventually so we use
-            -- @nf@ on the vector of 'IOOp's. We only evaluate the vector of
-            -- indexes to WHNF, because it is an unboxed vector.
+            -- prepLookups combines bloom filter querying and index searching.
+            -- The implementation forces the results to WHNF, so we use
+            -- whnfAppIO here instead of nfAppIO.
           , bench "Lookup preparation in memory" $
               whnfAppIO (\ks' -> prepLookups blooms indexes kopsFiles ks') ks
+            -- Submit the IOOps we get from prepLookups to HasBlockIO. We use
+            -- perRunEnv because IOOps contain mutable buffers, so we want fresh
+            -- ones for each run of the benchmark. We manually evaluate the
+            -- result to WHNF since it is unboxed vector.
+          , bench "Submit IOOps in batches" $
+              perRunEnv (prepLookups blooms indexes kopsFiles ks) $ \ ~(_rkixs, ioops) -> do
+                !_ioress <- submitInBatches hasBlockIO bsize ioops
+                pure ()
+            -- When IO result have been collected, intra-page lookups searches
+            -- through the raw bytes (representing a disk page) for the lookup
+            -- key. Again, we use perRunEnv here because IOOps contain mutable
+            -- buffers, so we want fresh ones for each run of the benchmark. The
+            -- result is a boxed vector of Maybe Entry, but since the
+            -- implementation takes care to evaluate each of the elements, we
+            -- only compute WHNF.
+          , bench "Perform intra-page lookups" $
+              perRunEnv ( prepLookups blooms indexes kopsFiles ks >>= \(rkixs, ioops) ->
+                          submitInBatches hasBlockIO bsize ioops >>= \ioress ->
+                          pure (rkixs, ioops, ioress)
+                        ) $ \ ~(rkixs, ioops, ioress) -> do
+                !_ <- intraPageLookups resolveV rs ks rkixs ioops ioress
+                pure ()
+            -- The whole shebang: lookup preparation, doing the IO in batches,
+            -- and then performing intra-page-lookups. Again, we evaluate the
+            -- result to WHNF because it is the same result that
+            -- intraPageLookups produces (see above).
           , bench "Batched lookups in IO" $
-              nfAppIO (\ks' -> lookupsInBatches hasBlockIO bsize resolveV rs blooms indexes kopsFiles ks') ks
+              whnfAppIO (\ks' -> lookupsInBatches hasBlockIO bsize resolveV rs blooms indexes kopsFiles ks') ks
```

Comment:
Did this make a noticeable difference?

Reply:
The difference is not very noticeable, or at least small enough that I can't say for certain whether there is a significant difference. We're talking about a ~10 micro-seconds difference in the positive lookup scenario, and no real difference in the negative lookup scenario

Reply:
Conceptually `whnfAppIO` is the right thing to do: the data should already be fully evaluated